[AWS] Add Redshift Integration

girodav opened the pull request:
<!-- Type of change
Please label this PR with one of the following labels, depending on the scope of your change:
- Bug
- Enhancement
- Breaking change
- Deprecation
-->

## What does this PR do?

<!-- Mandatory
Explain here the changes you made on the PR.
-->

This PR adds a new datastream  for [Amazon Redshift](https://docs.aws.amazon.com/redshift/latest/mgmt/welcome.html) metrics.

- Add new datastream to collect Redshift metrics through Cloudwatch
- Add system tests
- Bump AWS package version and change changelog accordingly
- Add dashboard and screenshots

## Checklist

- [x] I have reviewed [tips for building integrations](https://github.com/elastic/integrations/blob/main/docs/tips_for_building_integrations.md) and this pull request is aligned with them.
- [x] I have verified that all data streams collect metrics or logs.
- [x] I have added an entry to my package's `changelog.yml` file.
- [x] I have verified that Kibana version constraints are current according to [guidelines](https://github.com/elastic/elastic-package/blob/master/docs/howto/stack_version_support.md#when-to-update-the-condition).

## How to test this PR locally

```elastic-package test system --data-streams redshift```


## Related issues

- Closes https://github.com/elastic/integrations/issues/2902

## Screenshots

![Screenshot 2022-06-23 at 09 28 50](https://user-images.githubusercontent.com/1390902/175262176-8f4f6f69-994c-4002-9860-3748e4b1c233.png)

![metricbeat-aws-redshift-overview](https://user-images.githubusercontent.com/1390902/175262133-79acbade-d1e2-4088-b32e-704202184388.png)

![Screenshot 2022-06-27 at 16 43 57](https://user-images.githubusercontent.com/1390902/175980093-ecc9e629-1633-440e-a154-2d5cde5edb73.png)

elasticmachine commented:
## :broken_heart: Build Failed


<!-- BUILD BADGES-->
> _the below badges are clickable and redirect to their specific view in the CI or DOCS_
[![Pipeline View](https://img.shields.io/badge/pipeline-pipeline%20-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-3569/detail/PR-3569/1//pipeline) [![Test View](https://img.shields.io/badge/test-test-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-3569/detail/PR-3569/1//tests) [![Changes](https://img.shields.io/badge/changes-changes-green)](https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-3569/detail/PR-3569/1//changes) [![Artifacts](https://img.shields.io/badge/artifacts-artifacts-yellow)](https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-3569/detail/PR-3569/1//artifacts) [![preview](https://img.shields.io/badge/docs-preview-yellowgreen)](http://integrations_3569.docs-preview.app.elstc.co/diff)  [![preview](https://img.shields.io/badge/elastic-observability-blue)](https://ci-stats.elastic.co/app/apm/services/beats-ci/transactions/view?rangeFrom=2022-06-23T08:51:54.183Z&rangeTo=2022-06-23T09:11:54.183Z&transactionName=Ingest-manager/integrations/PR-{number}&transactionType=job&latencyAggregationType=avg&traceId=7292a4dd97b82ecc2d1c2c3aaf1dd317&transactionId=0d19bc49c82b82ed)

<!-- BUILD SUMMARY-->
<details><summary>Expand to view the summary</summary>
<p>

#### Build stats




* Start Time: 2022-06-23T09:01:54.183+0000


* Duration: 20 min 24 sec





</p>
</details>

<!-- TEST RESULTS IF ANY-->


<!-- STEPS ERRORS IF ANY -->












### Steps errors [![2](https://img.shields.io/badge/2%20-red)](https://beats-ci.elastic.co/blue/organizations/jenkins/Ingest-manager%2Fintegrations%2FPR-3569/detail/PR-3569/1//pipeline)
  <details><summary>Expand to view the steps failures</summary>
  <p>


  
  
  
  
##### `Test integration: aws`
<ul>
<li>Took 4 min 16 sec . View more details <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Ingest-manager/pipelines/integrations/pipelines/PR-3569/runs/1/steps/870/log/?start=0">here</a></li>
<li>Description: <code>                                eval "$(../../build/elastic-package stack shellinit)"                                ../../build/elastic-package test -v --report-format xUnit --report-output file --test-coverage                                </code></l1>
</ul>
  
  
  
  
##### `Google Storage Download`
<ul>
<li>Took 0 min 1 sec . View more details <a href="https://beats-ci.elastic.co//blue/rest/organizations/jenkins/pipelines/Ingest-manager/pipelines/integrations/pipelines/PR-3569/runs/1/steps/954/log/?start=0">here</a></li>

</ul>
  
  </p>
  </details>



## :robot: GitHub comments

To re-run your PR in the CI, just comment with:



- `/test` : Re-trigger the build.




<!--COMMENT_GENERATED_WITH_ID_comment.id-->

elasticmachine commented:
### :globe_with_meridians: Coverage report

Name | Metrics % (`covered/total`) | Diff
 --- | --- | ---
Packages | 100.0% (`11/11`) | :green_heart: 
Files | 91.667% (`11/12`) | :-1: -4.77
Classes | 91.667% (`11/12`) | :-1: -4.77
Methods | 82.692% (`172/208`) | :-1: -6.711
Lines | 92.673% (`2087/2252`) | :+1: 1.78
Conditionals | 100.0% (`0/0`) | :green_heart: 

<!--COMMENT_GENERATED_WITH_ID_coverage-->

kaiyan-sheng commented on packages/aws/data_stream/redshift/elasticsearch/ingest_pipeline/default.yml:
```diff
@@ -0,0 +1,143 @@
+      field: aws.dimensions.service class
+      target_field: aws.dimensions.service_class
+      ignore_missing: true
+  - rename:
```
We ususally don't rename these fields from CloudWatch. Instead, we keep them in the format of `aws.<namespace>.metrics.<metricname>.<stat>`. The reason why you see some of the data streams in AWS integration with renaming is to either keep backward compatibility or migrate to use ECS field names.

ravikesarwani commented:
cc: @Udayel Can you review?

@girodav This is great work. 
Can you please provide images for "Add integration" screens as well?
I would like to take a quick peek at the integration docs as well.

I am off on vacation for a little bit so please don't wait on me for merge.

kaiyan-sheng commented on packages/aws/data_stream/redshift/agent/stream/stream.yml.hbs:
```diff
@@ -0,0 +1,36 @@
+{{/if}}
+metrics:
+- namespace: "AWS/Redshift"
+  statistic: ["Average"]
```
Some metrics like `HealthStatus`, `MaintainanceMode` probably doesn't fit with the average statistic.

girodav replied:
We could potentially use "Minimum" for those 2, but I am not 100% sure if it would make a lot of difference. Redshift publishes 1 if the cluster is healthy, and 0 if not. There might be values in between, due to sampling from their side. Due to this, any value lower than 1 has to be considered "unhealthy" ([docs](https://docs.aws.amazon.com/redshift/latest/mgmt/metrics-listing.html)). We could incur in issues only if a floating point value is rounded up to 1.

Average is also what shown by default on the Redshift console for metrics (including these 2)

kaiyan-sheng reviewed (approved):
LGTM! Good work!

zmoog commented on packages/aws/data_stream/redshift/fields/agent.yml:
```diff
@@ -0,0 +1,198 @@
+- name: cloud
+  title: Cloud
+  group: 2
+  description: Fields related to the cloud or infrastructure the events are coming from.
+  footnote: 'Examples: If Metricbeat is running on an EC2 host and fetches data from its host, the cloud info contains the data about this machine. If Metricbeat runs on a remote machine outside the cloud and fetches data from a service running in the cloud, the field contains cloud data from the machine the service is running on.'
+  type: group
+  fields:
+    - name: account.id
+      level: extended
+      type: keyword
+      ignore_above: 1024
+      description: 'The cloud account or organization id used to identify different entities in a multi-tenant environment.
+
+        Examples: AWS account id, Google Cloud ORG Id, or other unique identifier.'
+      example: 666777888999
```
Do we need to add `cloud.account.id` here when it exists in the `ecs.yml` file as external?

zmoog commented on packages/aws/data_stream/redshift/fields/agent.yml:
```diff
@@ -0,0 +1,198 @@
+    - name: image.id
+      type: keyword
+      description: Image ID for the cloud instance.
+- name: container
```
Can we leverage these fields' [ECS definition](https://github.com/elastic/ecs/blob/main/schemas/container.yml)?

zmoog reviewed (commented):
This is excellent work @girodav ! 

I only have a few questions about ECS fields. I'm not 100% how it should work: should we leverage ECS for fields definitions for groups like `cloud.*`, `container.*`, and `host.*` vs. adding definition in the `.yml` file in the integration? 

Maybe @kaiyan-sheng can give some advice! 😇

girodav commented:
@kaiyan-sheng @zmoog As mentioned to you offline, I added ```agents.yml``` as I honestly thought it was needed as it is defined by other integrations. I 100% agree in using ECS fields only, if possible. I added this change in the latest commit

cachedout commented:
@girodav There are a few Redshift clusters that seem to have been created as a part of the automated tests introduced here which were never cleaned up:

https://us-east-1.console.aws.amazon.com/redshiftv2/home?region=us-east-1#clusters

Could you please see if those can be removed? Thanks.
